File: services/api/internal/domain/media/schedule_slot.go
package media

import (
	"errors"
	"time"

	"github.com/kurasikapa/api/internal/domain/identity"
	"github.com/kurasikapa/api/internal/domain/shared"
)

type ScheduleState string

const (
	ScheduleScheduled ScheduleState = "scheduled"
	ScheduleCancelled ScheduleState = "cancelled"
	ScheduleCompleted ScheduleState = "completed"
)

var (
	ErrScheduleInPast         = errors.New("schedule slot must start in the future")
	ErrInvalidScheduleWindow  = errors.New("schedule slot must end after it starts")
	ErrSlotAlreadyCancelled   = errors.New("schedule slot already cancelled")
	ErrReplayNeedsCaptions    = errors.New("recorded replay requires captions")
	ErrReplayBeforeSlotEnds   = errors.New("recorded replay cannot publish before the slot ends")
	ErrReplayAlreadyPublished = errors.New("recorded replay is already published")
	ErrReplayNotAwaiting      = errors.New("schedule slot is not awaiting a replay")
)

type ScheduleSlotState struct {
	ID                            shared.ScheduleSlotID
	ProgrammeID                   shared.ProgrammeID
	Locale                        string
	StartsAt, EndsAt              time.Time
	IsLive                        bool
	State                         ScheduleState
	ReplayAssetID, CaptionAssetID *shared.AssetID
	CreatedBy                     shared.UserID
}
type ScheduleSlot struct{ state ScheduleSlotState }

func NewScheduleSlot(actor identity.Actor, state ScheduleSlotState, now time.Time) (ScheduleSlot, error) {
	if err := actor.Require(identity.PermStreamManage); err != nil {
		return ScheduleSlot{}, err
	}
	if !state.StartsAt.After(now) {
		return ScheduleSlot{}, ErrScheduleInPast
	}
	if !state.EndsAt.After(state.StartsAt) {
		return ScheduleSlot{}, ErrInvalidScheduleWindow
	}
	state.State, state.CreatedBy = ScheduleScheduled, actor.ID()
	state.ReplayAssetID, state.CaptionAssetID = nil, nil
	return ScheduleSlot{state: state}, nil
}
func ReconstituteScheduleSlot(state ScheduleSlotState) ScheduleSlot {
	return ScheduleSlot{state: state}
}
func (s ScheduleSlot) State() ScheduleSlotState  { return s.state }
func (s ScheduleSlot) ID() shared.ScheduleSlotID { return s.state.ID }
func (s ScheduleSlot) Cancel(actor identity.Actor) (ScheduleSlot, error) {
	if err := actor.Require(identity.PermStreamManage); err != nil {
		return ScheduleSlot{}, err
	}
	if s.state.State == ScheduleCancelled {
		return ScheduleSlot{}, ErrSlotAlreadyCancelled
	}
	s.state.State = ScheduleCancelled
	return s, nil
}
func (s ScheduleSlot) PublishReplay(actor identity.Actor, replay shared.AssetID, captions *shared.AssetID, now time.Time) (ScheduleSlot, error) {
	if err := actor.Require(identity.PermStreamManage); err != nil {
		return ScheduleSlot{}, err
	}
	if s.state.State == ScheduleCompleted {
		return ScheduleSlot{}, ErrReplayAlreadyPublished
	}
	if s.state.State != ScheduleScheduled || !s.state.IsLive {
		return ScheduleSlot{}, ErrReplayNotAwaiting
	}
	if now.Before(s.state.EndsAt) {
		return ScheduleSlot{}, ErrReplayBeforeSlotEnds
	}
	if captions == nil {
		return ScheduleSlot{}, ErrReplayNeedsCaptions
	}
	s.state.State, s.state.ReplayAssetID, s.state.CaptionAssetID = ScheduleCompleted, &replay, captions
	return s, nil
}

File: packages/domain/src/media/schedule-slot.ts
import { type Actor, requirePermission } from '../identity/actor'
import type { AssetId, ProgrammeId, ScheduleSlotId, UserId } from '../shared/ids'

export type ScheduleSlotState = 'scheduled' | 'cancelled' | 'completed'

export interface ScheduleSlotProps {
  readonly id: ScheduleSlotId
  readonly programmeId: ProgrammeId
  readonly locale: string
  readonly startsAt: Date
  readonly endsAt: Date
  readonly isLive: boolean
  readonly state: ScheduleSlotState
  readonly replayAssetId: AssetId | null
  readonly captionAssetId: AssetId | null
  readonly createdBy: UserId
}

export type NewScheduleSlot = Omit<
  ScheduleSlotProps,
  'state' | 'replayAssetId' | 'captionAssetId' | 'createdBy'
>

export class ScheduleSlotInPast extends Error {}
export class InvalidScheduleWindow extends Error {}
export class SlotAlreadyCancelled extends Error {}
export class ReplayNeedsCaptions extends Error {}

/** One public airing of a recurring programme, optionally linked to its replay. */
export class ScheduleSlot {
  private constructor(private readonly props: ScheduleSlotProps) {}

  static schedule(actor: Actor, input: NewScheduleSlot, now: Date): ScheduleSlot {
    requirePermission(actor, 'stream:manage')
    if (input.startsAt.getTime() <= now.getTime()) throw new ScheduleSlotInPast()
    if (input.endsAt.getTime() <= input.startsAt.getTime()) throw new InvalidScheduleWindow()

    return new ScheduleSlot({
      ...input,
      state: 'scheduled',
      replayAssetId: null,
      captionAssetId: null,
      createdBy: actor.id,
    })
  }

  static reconstitute(props: ScheduleSlotProps): ScheduleSlot {
    return new ScheduleSlot(props)
  }

  get id(): ScheduleSlotId { return this.props.id }
  get programmeId(): ProgrammeId { return this.props.programmeId }
  get locale(): string { return this.props.locale }
  get startsAt(): Date { return this.props.startsAt }
  get endsAt(): Date { return this.props.endsAt }
  get isLive(): boolean { return this.props.isLive }
  get state(): ScheduleSlotState { return this.props.state }
  get replayAssetId(): AssetId | null { return this.props.replayAssetId }
  get captionAssetId(): AssetId | null { return this.props.captionAssetId }

  cancel(actor: Actor): ScheduleSlot {
    requirePermission(actor, 'stream:manage')
    if (this.props.state === 'cancelled') throw new SlotAlreadyCancelled()
    return this.with({ state: 'cancelled' })
  }

  publishReplay(actor: Actor, replayAssetId: AssetId, captionAssetId: AssetId | null): ScheduleSlot {
    requirePermission(actor, 'stream:manage')
    if (captionAssetId === null) throw new ReplayNeedsCaptions()
    return this.with({ state: 'completed', replayAssetId, captionAssetId })
  }

  snapshot(): ScheduleSlotProps {
    return this.props
  }

  private with(patch: Partial<ScheduleSlotProps>): ScheduleSlot {
    return new ScheduleSlot({ ...this.props, ...patch })
  }
}
